### DISCUSSION TITLE
Update to 1.25.0 fails

### DISCUSSION THREAD

**Discussion by angelbarrera92** [dani-garcia/vaultwarden#2519] [Q&A] [answered]
### Subject of the issue
Running vaultwarden 1.24.0, then, trying to update to 1.25.0 (just changing the container tag image). It fails during container startup. 

Seems like the migration sql sentences are not working fine in my setup.

### Deployment environment

* Vaultwarden version: v1.24.0
* Web-vault version: v2.25.1
* Running within Docker: false (Base: Debian)
* Environment settings overridden: true
* Uses a reverse proxy: true
* IP Header check: true (X-Real-IP)
* Internet access: true
* Internet access via a proxy: false
* DNS Check: true
* Time Check: true
* Domain Configuration Check: true
* HTTPS Check: true
* Database type: SQLite
* Database version: 3.35.4
* Clients used: 
* Reverse proxy and version: 
* Other relevant information: 

### Config (Generated via diagnostics page)
<details><summary>Show Running Config</summary>

**Environment settings which are overridden:** DOMAIN, SIGNUPS_ALLOWED, ADMIN_TOKEN, SMTP_HOST, SMTP_SSL, SMTP_EXPLICIT_TLS, SMTP_PORT, SMTP_FROM, SMTP_FROM_NAME, SMTP_USERNAME, SMTP_PASSWORD, SMTP_AUTH_MECHANISM

```json
{
  "_duo_akey": null,
  "_enable_duo": false,
  "_enable_email_2fa": false,
  "_enable_smtp": true,
  "_enable_yubico": true,
  "_ip_header_enabled": true,
  "admin_ratelimit_max_burst": 3,
  "admin_ratelimit_seconds": 300,
  "admin_token": "***",
  "allowed_iframe_ancestors": "",
  "attachments_folder": "/mnt/data/attachments",
  "authenticator_disable_time_drift": false,
  "data_folder": "/mnt/data",
  "database_max_conns": 10,
  "database_url": "/***/****/**.*******",
  "db_connection_retries": 15,
  "disable_2fa_remember": false,
  "disable_admin_token": false,
  "disable_icon_download": false,
  "domain": "*****://*****.*.**********.***",
  "domain_origin": "*****://*****.*.**********.***",
  "domain_path": "",
  "domain_set": true,
  "duo_host": null,
  "duo_ikey": null,
  "duo_skey": null,
  "email_attempts_limit": 3,
  "email_expiration_time": 600,
  "email_token_size": 6,
  "emergency_access_allowed": true,
  "emergency_notification_reminder_schedule": "0 5 * * * *",
  "emergency_request_timeout_schedule": "0 5 * * * *",
  "enable_db_wal": true,
  "extended_logging": true,
  "helo_name": null,
  "hibp_api_key": null,
  "icon_blacklist_non_global_ips": true,
  "icon_blacklist_regex": null,
  "icon_cache_folder": "/mnt/data/icon_cache",
  "icon_cache_negttl": 259200,
  "icon_cache_ttl": 2592000,
  "icon_download_timeout": 10,
  "icon_redirect_code": 302,
  "icon_service": "internal",
  "incomplete_2fa_schedule": "30 * * * * *",
  "incomplete_2fa_time_limit": 3,
  "invitation_org_name": "Bitwarden_RS",
  "invitations_allowed": true,
  "ip_header": "X-Real-IP",
  "job_poll_interval_ms": 30000,
  "log_file": null,
  "log_level": "Info",
  "log_timestamp_format": "%Y-%m-%d %H:%M:%S.%3f",
  "login_ratelimit_max_burst": 10,
  "login_ratelimit_seconds": 60,
  "org_attachment_limit": null,
  "org_creation_users": "",
  "password_iterations": 100000,
  "reload_templates": false,
  "require_device_email": false,
  "rsa_key_filename": "/mnt/data/rsa_key",
  "send_purge_schedule": "0 5 * * * *",
  "sends_allowed": true,
  "sends_folder": "/mnt/data/sends",
  "show_password_hint": true,
  "signups_allowed": false,
  "signups_domains_whitelist": "",
  "signups_verify": false,
  "signups_verify_resend_limit": 6,
  "signups_verify_resend_time": 3600,
  "smtp_accept_invalid_certs": false,
  "smtp_accept_invalid_hostnames": false,
  "smtp_auth_mechanism": "Login",
  "smtp_debug": false,
  "smtp_explicit_tls": true,
  "smtp_from": "***********@**********.***",
  "smtp_from_name": "Vaultwarden",
  "smtp_host": "****.**.*******.***",
  "smtp_password": "***",
  "smtp_port": 465,
  "smtp_ssl": true,
  "smtp_timeout": 15,
  "smtp_username": "***********@*.**********.***",
  "templates_folder": "/mnt/data/templates",
  "trash_auto_delete_days": null,
  "trash_purge_schedule": "0 5 0 * * *",
  "use_syslog": false,
  "user_attachment_limit": null,
  "web_vault_enabled": true,
  "web_vault_folder": "web-vault/",
  "websocket_address": "0.0.0.0",
  "websocket_enabled": true,
  "websocket_port": 3012,
  "yubico_client_id": null,
  "yubico_secret_key": null,
  "yubico_server": null
}
```
</details>

### Steps to reproduce

I'm updating stopping the 1.24.0 and then starting 1.25.0 with the same configuration.

### Expected behaviour

It updates :)

### Actual behaviour

It fails with the following error:

```bash
Running migration 20220302210038
Executing migration script 20220302210038/up.sql
[2022-05-30 06:42:18.281][vaultwarden::util][WARN] Can't connect to database, retrying: DieselMig.
[CAUSE] QueryError(
    DatabaseError(
        __Unknown,
        "NOT NULL constraint failed: devices_new.updated_at",
    ),
)
```

### Troubleshooting data

Non-default config:

```
  DATA_FOLDER: /mnt/data
  DOMAIN: "https://vault.<REDACTED>"
  WEBSOCKET_ENABLED: "true"
  ADMIN_TOKEN: "<REDACTED>"
  SIGNUPS_ALLOWED: "false"
  ADMIN_BASIC_AUTH: "<REDACTED>:<REDACTED>"
  SMTP_HOST: <REDACTED>
  SMTP_PORT: "465"
  SMTP_SSL: "true"
  SMTP_EXPLICIT_TLS: "true"
  SMTP_USERNAME: "vaultwarden@<REDACTED>"
  SMTP_FROM: "vaultwarden@<REDACTED>"
  SMTP_FROM_NAME: "Vaultwarden"
  SMTP_PASSWORD: "<REDACTED>"
  SMTP_AUTH_MECHANISM: "Login"
```

1.25.0 start logs

```bash
/--------------------------------------------------------------------\
|                        Starting Vaultwarden                        |
|                           Version 1.25.0                           |
|--------------------------------------------------------------------|
| This is an *unofficial* Bitwarden implementation, DO NOT use the   |
| official channels to report bugs/features, regardless of client.   |
| Send usage/configuration questions or feature requests to:         |
|   https://vaultwarden.discourse.group/                             |
| Report suspected bugs/issues in the software itself at:            |
|   https://github.com/dani-garcia/vaultwarden/issues/new            |
\--------------------------------------------------------------------/

[INFO] No .env file found.

[WARNING] The following environment variables are being overriden by the config file,
[WARNING] please use the admin panel to make changes to them:
[WARNING] DOMAIN, SIGNUPS_ALLOWED, ADMIN_TOKEN, SMTP_HOST, SMTP_SSL, SMTP_EXPLICIT_TLS, SMTP_PORT, SMTP_FROM, SMTP_FROM_NAME, SMTP_USERNAME, SMTP_PASSWORD, SMTP_AUTH_MECHANISM

[DEPRECATED]: `SMTP_SSL` or `SMTP_EXPLICIT_TLS` is set. Please use `SMTP_SECURITY` instead.
Running migration 20220302210038
Executing migration script 20220302210038/up.sql
[2022-05-30 06:42:18.281][vaultwarden::util][WARN] Can't connect to database, retrying: DieselMig.
[CAUSE] QueryError(
    DatabaseError(
        __Unknown,
        "NOT NULL constraint failed: devices_new.updated_at",
    ),
)
Running migration 20220302210038
Executing migration script 20220302210038/up.sql
[2022-05-30 06:42:19.291][vaultwarden::util][WARN] Can't connect to database, retrying: DieselMig.
[CAUSE] QueryError(
    DatabaseError(
        __Unknown,
        "NOT NULL constraint failed: devices_new.updated_at",
    ),
)
Running migration 20220302210038
Executing migration script 20220302210038/up.sql
[2022-05-30 06:42:20.300][vaultwarden::util][WARN] Can't connect to database, retrying: DieselMig.
[CAUSE] QueryError(
    DatabaseError(
        __Unknown,
        "NOT NULL constraint failed: devices_new.updated_at",
    ),
)
Running migration 20220302210038
Executing migration script 20220302210038/up.sql
[2022-05-30 06:42:21.309][vaultwarden::util][WARN] Can't connect to database, retrying: DieselMig.
[CAUSE] QueryError(
    DatabaseError(
        __Unknown,
        "NOT NULL constraint failed: devices_new.updated_at",
    ),
)
Running migration 20220302210038
Executing migration script 20220302210038/up.sql
[2022-05-30 06:42:22.321][vaultwarden::util][WARN] Can't connect to database, retrying: DieselMig.
[CAUSE] QueryError(
    DatabaseError(
        __Unknown,
        "NOT NULL constraint failed: devices_new.updated_at",
    ),
)
Running migration 20220302210038
Executing migration script 20220302210038/up.sql
[2022-05-30 06:42:23.333][vaultwarden::util][WARN] Can't connect to database, retrying: DieselMig.
[CAUSE] QueryError(
    DatabaseError(
        __Unknown,
        "NOT NULL constraint failed: devices_new.updated_at",
    ),
)
Running migration 20220302210038
Executing migration script 20220302210038/up.sql
[2022-05-30 06:42:24.343][vaultwarden::util][WARN] Can't connect to database, retrying: DieselMig.
[CAUSE] QueryError(
    DatabaseError(
        __Unknown,
        "NOT NULL constraint failed: devices_new.updated_at",
    ),
)
Running migration 20220302210038
Executing migration script 20220302210038/up.sql
[2022-05-30 06:42:25.354][vaultwarden::util][WARN] Can't connect to database, retrying: DieselMig.
[CAUSE] QueryError(
    DatabaseError(
        __Unknown,
        "NOT NULL constraint failed: devices_new.updated_at",
    ),
)
Running migration 20220302210038
Executing migration script 20220302210038/up.sql
[2022-05-30 06:42:26.375][vaultwarden::util][WARN] Can't connect to database, retrying: DieselMig.
[CAUSE] QueryError(
    DatabaseError(
        __Unknown,
        "NOT NULL constraint failed: devices_new.updated_at",
    ),
)
Running migration 20220302210038
Executing migration script 20220302210038/up.sql
[2022-05-30 06:42:27.385][vaultwarden::util][WARN] Can't connect to database, retrying: DieselMig.
[CAUSE] QueryError(
    DatabaseError(
        __Unknown,
        "NOT NULL constraint failed: devices_new.updated_at",
    ),
)
```

**Comment by BlackDex** [comment]
Strange, there seems to be something invalid already with your current `devices` table.
Best solution here would be to stop Vaultwarden, make a copy of the database as backup.
Then open the database using a sqlite editor (GUI is the most easy for this one).
And then look at the devices table which has nothing filled in at the `updated_at` table.
Then you could add a custom date there and try to rerun the the upgrade.

An other option would be (if that works) is to remove all sessions for all users via the admin interface.
That should clear that table, but also forces  you to do a re-login on all devices for all users.

**Comment by angelbarrera92** [comment]
Added a dummy date in those row where were null; new errors appear in the same table while re-running the upgrade:
```
Running migration 20220302210038
Executing migration script 20220302210038/up.sql
[2022-05-30 15:22:23.227][vaultwarden][ERROR] Error creating database pool: DieselMig.
[CAUSE] QueryError(
    DatabaseError(
        __Unknown,
        "NOT NULL constraint failed: devices_new.user_uuid",
    ),
)
```

**Comment by BlackDex** [comment]
Well, there are just two fields/columns which may be empty, and those are `push_token` and `twofactor_remember` all other fields should be filled.
If you have a records which do not have a `user_uuid` i would remove those records and try again

**Comment by angelbarrera92** [comment]
No luck...
Well, at least something changed... I run `delete from devices where users_uuid is null`, then re-run the upgrade, but no passwords appear. 

Logs reports: 

```
[2022-05-30 15:38:04.864][panic][ERROR] thread 'rocket-worker-thread' panicked at 'Error loading ciphers: DatabaseError(__Unknown, "database disk image is malformed")': src/db/models/cipher.rs:518
   0: vaultwarden::init_logging::{{closure}}
   1: std::panicking::rust_panic_with_hook
   2: std::panicking::begin_panic_handler::{{closure}}
   3: std::sys_common::backtrace::__rust_end_short_backtrace
   4: rust_begin_unwind
   5: core::panicking::panic_fmt
   6: core::result::unwrap_failed
   7: tokio::runtime::enter::exit
   8: tokio::runtime::thread_pool::worker::block_in_place
   9: <core::future::from_generator::GenFuture<T> as core::future::future::Future>::poll
  10: <core::future::from_generator::GenFuture<T> as core::future::future::Future>::poll
  11: <core::future::from_generator::GenFuture<T> as core::future::future::Future>::poll
  12: <core::future::from_generator::GenFuture<T> as core::future::future::Future>::poll
  13: tokio::runtime::task::harness::poll_future
  14: tokio::runtime::task::harness::Harness<T,S>::poll
  15: std::thread::local::LocalKey<T>::with
  16: tokio::runtime::thread_pool::worker::Context::run_task
  17: tokio::runtime::thread_pool::worker::Context::run
  18: tokio::macros::scoped_tls::ScopedKey<T>::set
  19: tokio::runtime::thread_pool::worker::run
  20: <tokio::runtime::blocking::task::BlockingTask<T> as core::future::future::Future>::poll
  21: tokio::runtime::task::harness::Harness<T,S>::poll
  22: tokio::runtime::blocking::pool::Inner::run
  23: std::sys_common::backtrace::__rust_begin_short_backtrace
  24: core::ops::function::FnOnce::call_once{{vtable.shim}}
  25: std::sys::unix::thread::Thread::new::thread_start
  26: start_thread
  27: clone

[2022-05-30 15:38:04.865][response][INFO] (sync) GET /api/sync?<data..> => 500 Internal Server Error
```

**Comment by BlackDex** [comment]
Now it looks like the database is broken.
Maybe something went wrong with the query, or the database is not closed yet from all processes.

**Reply by BlackDex** [reply]
I would suggest to take steps 2b from this link https://github.com/dani-garcia/vaultwarden/discussions/2511#discussioncomment-2841448

**Reply by angelbarrera92** [reply]
I can not make it work... The database is a sqlite3 on an NFS mount.
The latest test was in the same filesystem as the vault process (stopped), (before I downloaded the SQLite file to my local pc, modify, then re-upload). 

Starting in 1.24 with the modified database (the delete statement):

```bash
[2022-05-30 15:55:14.381][panic][ERROR] thread 'unnamed' panicked at 'Error loading ciphers: DatabaseError(__Unknown, "database disk image is malformed")': src/db/models/cipher.rs:470
   0: vaultwarden::init_logging::{{closure}}
   1: std::panicking::rust_panic_with_hook
             at rustc/bfe15646761a75f0259e204cab071565eed2b1e5/library/std/src/panicking.rs:695:17
   2: std::panicking::begin_panic_handler::{{closure}}
             at rustc/bfe15646761a75f0259e204cab071565eed2b1e5/library/std/src/panicking.rs:581:13
   3: std::sys_common::backtrace::__rust_end_short_backtrace
             at rustc/bfe15646761a75f0259e204cab071565eed2b1e5/library/std/src/sys_common/backtrace.rs:139:18
   4: rust_begin_unwind
             at rustc/bfe15646761a75f0259e204cab071565eed2b1e5/library/std/src/panicking.rs:577:5
   5: core::panicking::panic_fmt
             at rustc/bfe15646761a75f0259e204cab071565eed2b1e5/library/core/src/panicking.rs:135:14
   6: core::result::unwrap_failed
             at rustc/bfe15646761a75f0259e204cab071565eed2b1e5/library/core/src/result.rs:1737:5
   7: vaultwarden::db::models::cipher::Cipher::find_by_user_visible
   8: vaultwarden::api::core::ciphers::rocket_route_fn_sync
   9: <F as rocket::handler::Handler>::handle
  10: rocket::rocket::Rocket::route_and_process
  11: <rocket::rocket::Rocket as hyper::server::Handler>::handle
  12: hyper::server::Worker<H>::handle_connection
  13: std::sys_common::backtrace::__rust_begin_short_backtrace
  14: core::ops::function::FnOnce::call_once{{vtable.shim}}
  15: <alloc::boxed::Box<F,A> as core::ops::function::FnOnce<Args>>::call_once
             at rustc/bfe15646761a75f0259e204cab071565eed2b1e5/library/alloc/src/boxed.rs:1854:9
      <alloc::boxed::Box<F,A> as core::ops::function::FnOnce<Args>>::call_once
             at rustc/bfe15646761a75f0259e204cab071565eed2b1e5/library/alloc/src/boxed.rs:1854:9
      std::sys::unix::thread::Thread::new::thread_start
             at rustc/bfe15646761a75f0259e204cab071565eed2b1e5/library/std/src/sys/unix/thread.rs:108:17
  16: start_thread
  17: clone

[2022-05-30 15:55:14.382][panic][ERROR] thread 'unnamed' panicked at 'Error loading ciphers: DatabaseError(__Unknown, "database disk image is malformed")': src/db/models/cipher.rs:470
   0: vaultwarden::init_logging::{{closure}}
   1: std::panicking::rust_panic_with_hook
             at rustc/bfe15646761a75f0259e204cab071565eed2b1e5/library/std/src/panicking.rs:695:17
   2: std::panicking::begin_panic_handler::{{closure}}
             at rustc/bfe15646761a75f0259e204cab071565eed2b1e5/library/std/src/panicking.rs:581:13
   3: std::sys_common::backtrace::__rust_end_short_backtrace
             at rustc/bfe15646761a75f0259e204cab071565eed2b1e5/library/std/src/sys_common/backtrace.rs:139:18
   4: rust_begin_unwind
             at rustc/bfe15646761a75f0259e204cab071565eed2b1e5/library/std/src/panicking.rs:577:5
   5: core::panicking::panic_fmt
             at rustc/bfe15646761a75f0259e204cab071565eed2b1e5/library/core/src/panicking.rs:135:14
   6: core::result::unwrap_failed
             at rustc/bfe15646761a75f0259e204cab071565eed2b1e5/library/core/src/result.rs:1737:5
   7: vaultwarden::db::models::cipher::Cipher::find_by_user_visible
   8: vaultwarden::api::core::ciphers::rocket_route_fn_sync
   9: <F as rocket::handler::Handler>::handle
  10: rocket::rocket::Rocket::route_and_process
  11: <rocket::rocket::Rocket as hyper::server::Handler>::handle
  12: hyper::server::Worker<H>::handle_connection
  13: std::sys_common::backtrace::__rust_begin_short_backtrace
  14: core::ops::function::FnOnce::call_once{{vtable.shim}}
  15: <alloc::boxed::Box<F,A> as core::ops::function::FnOnce<Args>>::call_once
             at rustc/bfe15646761a75f0259e204cab071565eed2b1e5/library/alloc/src/boxed.rs:1854:9
      <alloc::boxed::Box<F,A> as core::ops::function::FnOnce<Args>>::call_once
             at rustc/bfe15646761a75f0259e204cab071565eed2b1e5/library/alloc/src/boxed.rs:1854:9
      std::sys::unix::thread::Thread::new::thread_start
             at rustc/bfe15646761a75f0259e204cab071565eed2b1e5/library/std/src/sys/unix/thread.rs:108:17
  16: start_thread
  17: clone
```

I've updated vaultwarden multiple times :) I'm super happy with the project, this is the first update that is not properly working.

**Reply by angelbarrera92** [reply]
With a working db: `db.sqlite3.replica.backup.play`:

```
# sqlite3 db.sqlite3.replica.backup.play  "PRAGMA integrity_check;"
*** in database main ***
On tree page 9 cell 0: 2nd reference to page 166
On tree page 9 cell 0: 2nd reference to page 165
On tree page 10 cell 0: 2nd reference to page 164
On tree page 10 cell 0: 2nd reference to page 163
Page 94 is never used
Page 95 is never used
Page 117 is never used
Page 137 is never used
Page 147 is never used
Page 161 is never used
NULL value in devices.updated_at
NULL value in devices.user_uuid
NULL value in devices.name
NULL value in devices.atype
NULL value in devices.refresh_token
row 1 missing from index sqlite_autoindex_devices_1
NULL value in devices.updated_at
NULL value in devices.user_uuid
NULL value in devices.name
NULL value in devices.atype
NULL value in devices.refresh_token
row 2 missing from index sqlite_autoindex_devices_1
NULL value in devices.updated_at
NULL value in devices.user_uuid
NULL value in devices.name
NULL value in devices.atype
NULL value in devices.refresh_token
row 3 missing from index sqlite_autoindex_devices_1
NULL value in devices.updated_at
NULL value in devices.user_uuid
NULL value in devices.name
NULL value in devices.atype
NULL value in devices.refresh_token
row 4 missing from index sqlite_autoindex_devices_1
NULL value in devices.updated_at
NULL value in devices.user_uuid
NULL value in devices.name
NULL value in devices.atype
NULL value in devices.refresh_token
row 5 missing from index sqlite_autoindex_devices_1
NULL value in devices.updated_at
NULL value in devices.user_uuid
NULL value in devices.name
NULL value in devices.atype
NULL value in devices.refresh_token
row 6 missing from index sqlite_autoindex_devices_1
NULL value in devices.updated_at
NULL value in devices.user_uuid
NULL value in devices.name
NULL value in devices.atype
NULL value in devices.refresh_token
row 7 missing from index sqlite_autoindex_devices_1
NULL value in devices.updated_at
NULL value in devices.user_uuid
NULL value in devices.name
NULL value in devices.atype
NULL value in devices.refresh_token
row 8 missing from index sqlite_autoindex_devices_1
NULL value in devices.updated_at
NULL value in devices.user_uuid
NULL value in devices.name
NULL value in devices.atype
NULL value in devices.refresh_token
row 9 missing from index sqlite_autoindex_devices_1
NULL value in devices.updated_at
NULL value in devices.user_uuid
NULL value in devices.name
NULL value in devices.atype
NULL value in devices.refresh_token
row 10 missing from index sqlite_autoindex_devices_1
NULL value in devices.updated_at
NULL value in devices.user_uuid
NULL value in devices.name
NULL value in devices.atype
NULL value in devices.refresh_token
row 11 missing from index sqlite_autoindex_devices_1
NULL value in devices.updated_at
NULL value in devices.user_uuid
NULL value in devices.name
NULL value in devices.atype
NULL value in devices.refresh_token
row 12 missing from index sqlite_autoindex_devices_1
NULL value in devices.updated_at
NULL value in devices.user_uuid
NULL value in devices.name
NULL value in devices.atype
NULL value in devices.refresh_token
row 13 missing from index sqlite_autoindex_devices_1
NULL value in devices.updated_at
NULL value in devices.user_uuid
NULL value in devices.name
NULL value in devices.atype
NULL value in devices.refresh_token
row 14 missing from index sqlite_autoindex_devices_1
NULL value in devices.updated_at
NULL value in devices.user_uuid
NULL value in devices.name
NULL value in devices.atype
NULL value in devices.refresh_token
row 15 missing from index sqlite_autoindex_devices_1
/mnt/data # sqlite3 db.sqlite3.replica.backup.play 
SQLite version 3.34.1 2021-01-20 14:10:07
Enter ".help" for usage hints.
sqlite> delete from devices where user_uuid is null;
sqlite> .exit
/mnt/data # sqlite3 db.sqlite3.replica.backup.play  "PRAGMA integrity_check;"
*** in database main ***
On tree page 9 cell 0: 2nd reference to page 166
On tree page 9 cell 0: 2nd reference to page 165
On tree page 10 cell 0: 2nd reference to page 164
On tree page 10 cell 0: 2nd reference to page 163
Page 94 is never used
Page 95 is never used
Page 117 is never used
Page 137 is never used
Page 147 is never used
Page 161 is never used
Error: database disk image is malformed
```
TL;DR: 

1. Check the integrity. OK
2. run the delete statement
3. Check the integrity KO

**Reply by angelbarrera92** [reply]
Ok.... finally I made it work...
Not sure if this is the best way... but here i go:

```bash
>sqlite3 db.sqlite3
sqlite>.output backup.db
sqlite>.dump
sqlite>.quit

>sqlite3 db.sqlite3_fixed.db
sqlite>.read backup.db
sqlite>.quit

>mv db.sqlite3_fixed.db db.sqlite3
```

> During the `.read backup.db` will throw some errors (regarding the `null` values I had in my devices table). Seems to be working after a quick test.

*source: https://blog.niklasottosson.com/databases/sqlite-check-integrity-and-fix-common-problems/*

**Reply by BlackDex** [reply]
Running a sqlite database file on a Network filesystem can cause issues. That is probably why the database had issues already, but you didn't noticed it before.

Depending on the NFS version it might not have supported file locking completely, which then in turn causes errors.

**Comment by angelbarrera92** [comment, marked as answer]
Ok, I was able to sort the issue. I'll try to summarize the command I run:

- `cp db.sqlite3 db.sqlite3.backup`: Creates a backup of the original DB
- `sqlite3 db.sqlite3`: Open the original DB
  - `delete from devices where user_uuid is null;`
  - `.mode insert`
  - `.output dump_all.sql`
  - `.dump`
  - `.exit`
- `cat dump_all.sql | grep -v TRANSACTION | grep -v ROLLBACK | grep -v COMMIT >dump_all_notrans.sql`: Source: https://support.storj.io/hc/en-us/articles/360029309111-How-to-fix-a-database-disk-image-is-malformed-
- `sqlite3 db.sqlite3.restored ".read dump_all_notrans.sql"`
- `sqlite3 db.sqlite3.restored "PRAGMA integrity_check;"`
  - `ok`

Then, finally: `cp db.sqlite3.restored db.sqlite3` and start the vaultwarden process. Test if you can create/modify any password/secure-note. If everything is fine, update to 1.25.0.

Now it is working fine in 1.25.0.
Thanks for the duck duck debugging session.

**Reply by BlackDex** [reply]
Thanks for sharing the solution!
